### PR TITLE
Drop jQuery as a dependency

### PR DESCRIPTION
**Pull request by liudangyi**
As discussed in rails/rails#25208, web APIs in browsers have developed a lot in past few years, making it possible for us to build UJS without jQuery. This PR primitively refactors the whole UJS based on native APIs. 
### Browsers compatibility

According to liudangyi/jquery-ujs#1, the new UJS should support IE10+ and all other modern browsers.
### Main changes
1. Organize the source code as an NPM package using `webpack`, rather than a single monolithic file. Related PR: #473.
2. Use many ES6 features, with `babel` as precompiler.
3. Use `eslint` instead of `jshint`.
4. No jQuery. The detailed solution is available here: liudangyi/jquery-ujs#1.
### API changes
1. "overrides" feature is not implemented yet, because I can hardly find any possible scenario.
2. "window.rails" instead of "$.rails".
3. Event system. Since we are using native events (CustomEvent) now, there are some changes in binding / triggering an event if using jQuery. Some examples:

``` javascript
// Add an event listener using jQuery
// before
$(form).bind('ajax:success', function(e, data, status, xhr) {
  // code
})
// after
$(form).bind('ajax:success', function(e) {
  var data = e.originalEvent.detail[0]
  var status = e.originalEvent.detail[1]
  var xhr = e.originalEvent.detail[2]
  // code
})

// Manually trigger an event
// before
$(link).trigger('click')
// after
$(link)[0].click()
// before
$(form).trigger('submit')
// after, because el.submit() does't fire the submit event!
if ($(form)[0].dispatchEvent(new Event('submit', { bubbles: true, cancelable: true })) !== false) {
  $(form)[0].submit()
}
```
### Existing issues
1. (liudangyi/jquery-ujs#6) Naming. UJS should have a new name with no "jQuery". Any idea?
2. (liudangyi/jquery-ujs#5) As "event system" in last section, maybe we should add some helper functions for binding / triggering events. A potential solution (for jQuery users) exists in [test/settings.js](https://github.com/liudangyi/jquery-ujs/blob/develop/test/public/test/settings.js#L66). But is it necessary or not?

### CODE REVIEW

**Comment by liudangyi** [comment]
Webkit engine embedded in phantomjs is too old to support

1. `new MouseEvent()`, a workaround of which is `document.createEvent('MouseEvents').initMouseEvent`.
2. Complex query selector such as `button[data-remote]:not([form]):not(form button)`, which I don't found any workaround yet. :(

**Comment by rafaelfranca** [comment]
First important thing, we should not use webpack. Most of the rails contributors are more used to sprockets so we should use it.

**Comment by liudangyi** [comment]
@rafaelfranca `webpack` is only used to build the project. Sprockets users would use the pre-built version of UJS as before, which is distributed in both Ruby gem and NPM package. Only advanced users may need to require source code directly using NPM package.

**Comment by lucasmazza** [comment]
The ES6 usage and inline docs are looking great!

I agree with @rafaelfranca on avoiding webpack and sticking with sprockets which is closer to Rails: we could use [javan/blade](https://github.com/javan/blade) to build and maybe upgrade our test suite to use sauce labs, as it seems good enough for something like this given how it's used to build actioncable and turbolinks.

**Comment by liudangyi** [comment]
In fact I discussed with @pixeltrix on this in liudangyi/jquery-ujs#2. An advantage of using webpack module is that it really understands JavaScript dependency, while blade only understands file dependency and we need to put all exported symbols in a global variable. 

And if we decide to use blade, I would suggest that we also move from ES6 to CoffeeScript.

**Comment by liudangyi** [comment]
So I want to write a little more about why I chose webpack and what it does here.

### Why JavaScript module system rather than Sprockets

The rationale is that we want to make a better organization of the source code, which actually has two steps: first is to split the single monolithic file into multiple modules, second is to package these modules into a single distributable file.

The biggest problem of sprockets is that it's not module-aware at all. It only concatenates all source files in a given order. If we want to export a function in a module and use it in another module, we must assign it to a global variable. This is not a big problem when writing application code, but causes a lot of inconvenience in library development. 

|                                 | Sprockets                              | JS module system                         |
| ------------------------------- | -------------------------------------- | ---------------------------------------- |
| Dependency                      | Specify dependency manually            | Tools will resolve dependency automatically |
| Call functions in other modules | Use global namespace as a bridge       | Just export and import                   |
| Static code analysis            | Linter hardly handles global variables | Better linter support                    |
| For developer                   |                                        | Easier to understand project structure   |

### Why webpack rather than RequireJS, SystemJS, Browserify, etc.

There are many choices but we only have one principle: least size overhead. And in fact we needn't dynamical module loading at all, so that RequireJS and SystemJS seem too heavy. Browserify and webpack fit both, then the test result shows that webpack will add an extra 218 bytes while browserify adds 498 bytes. So I chose webpack.

### What does webpack really do here

1. Calls babel to compile ES6 code into ES5 code, without any polyfills.
2. Analyzes dependency and works out an order for all source files.
3. Concatenates all source files together.
4. Provides `require` and `exports` function with an overhead of 218 bytes (minimized).
5. Minimizes if necessary.

Since we only use very limited functionality of webpack, mostly developers need not learn webpack at all.

### What need contributors do for development 

1. Setup:
   1. `bundle install`
   2. `npm install` for eslint, webpack and babel.
2. Develop: `rake develop` to start webpack and test server
3. Lint: `npm run lint`
4. Build for production: `rake webpack:build` or `npm run build`

**Comment by kaspth** [comment]
It seems Turbolinks manages to be just an npm module with Sprockets through [Blade](https://github.com/javan/blade) just fine: https://github.com/turbolinks/turbolinks#installation-using-npm

Blade is the defacto way for the Rails organization to bundle JS assets, we use it for Action Cable's assets as well. Thus, although it may be great, we shouldn't use WebPack. ✌️

**Comment by liudangyi** [comment]
@kaspth Blade provides [UMD](https://github.com/umdjs/umd) support but it still has all the disadvantages of Sprockets listed above.

Also, Blade doesn't fit very well here because UJS test requires a custom server, but Blade hasn't provided a way to mount custom routes yet.

**Comment by kaspth** [comment]
@javan you have a lot more experience with using Sprockets for this sort of thing than I do. Do you have any pointers on how we could use Blade here instead of going full bore WebPack? And while still having some form of module support?

**Comment by rafaelfranca** [comment]
If we want module support with Sprockets we can use https://github.com/Shopify/sprockets-commoner

**Comment by javan** [comment]
> Why JavaScript module system rather than Sprockets

First of all, I'm sympathetic to this desire. It's how most of JavaScript ecosystem outside of Rails operates. But until the Rails asset pipeline supports modules, I don't think using them is the right move. We made our dog food and should continue eating it. Personally, I think it's the caviar of dog food.

> The biggest problem of sprockets is that it's not module-aware at all. It only concatenates all source files in a given order. If we want to export a function in a module and use it in another module, we must assign it to a global variable.

[sprockets-export](https://github.com/javan/sprockets-export) uses a single object as a *namespace*, but only assigns it as a `window` global when the .js file is *not* consumed by a module loader.

So, using Action Cable as an example, if you're using a tool like webpack in your application you can `const ActionCable = require('action_cable')` and no global will be set. If you're using good 'ol Sprockets in your application then `//= require action_cable` will set `window.ActionCable`.

Functions, properties, etc. can be imported from that namespace so splitting code up into separate logical files is straightforward and natural. [Here's an example from Trix](https://github.com/basecamp/trix/blob/master/src/trix/models/document.coffee#L4-L6).

> Also, Blade doesn't fit very well here because UJS test requires a custom server, but Blade hasn't provided a way to mount custom routes yet.

Blade builds on top of Sprockets and you can directly access its Sprocket environment. Basically, it plays nice with Rack and should be easily adapted to this project's test server. See https://github.com/basecamp/trix/blob/master/config.ru for an example.

**Comment by javan** [comment]
I think https://github.com/turbolinks/turbolinks/tree/master/src/turbolinks is a gold star example to follow for structuring a JavaScript library The Rails Way™.

**Comment by liudangyi** [comment]
@rafaelfranca Thank you for your information. I didn't notice sprockets-commoner before and it seems very suitable here.

@javan Using a single object as a namespace is acceptable, but I'd like to try sprockets-commoner first since it may require less modification to existing code. Blade is a better choice for building and testing. I'll gradually update the test code and migrate to blade (current test code doesn't work with QUnit 1.20.0).

**Comment by guilleiguaran** [comment]
I would love if we can reach a middle ground between "JS way" and "Rails way", IMHO this can be done keeping the current code as is (without switching from ES6 to CS) but using Sprockets and Blade instead of WebPack and sprockets-commoner for modules.

**Comment by liudangyi** [comment]
@guilleiguaran I've tried sprockets-commoner and it's just great! It plays the webpack role perfectly with even smaller generated file. 

Now I only have one problem: how to integrate blade with current test framework. Working on rack now...

**Comment by liudangyi** [comment]
I think it's quite OK now. We are no longer using webpack. But babel and eslint still need to be installed using npm.

Does anyone have time to review this PR?

**Comment by guilleiguaran** [comment]
Replying to your first question:

 > Naming. UJS should have a new name with no "jQuery". Any idea?

I think `rails-ujs` is a good name for this, opinions?

**Comment by lucasmazza** [comment]
> But babel and eslint still need to be installed using npm.

We can't get away from this, but as the only required step for development is `npm install`, I think it won't be a big overhead on maintainers/contributors as webpack could be. I'll see if I can do a review of the implementation only instead of the tooling (which is what we have been talking about so far) 

> I think rails-ujs is a good name for this, opinions?

:+1:, but should it replace the `jquery-ujs` repo or have a new one for this vanilla js implementation?

**Comment by javan** [comment]
I don't want to derail progress on the move to drop jQuery, but these changes are still vastly different from the other Rails JavaScript libraries like [Action Cable](https://github.com/rails/rails/tree/master/actioncable/app/assets/javascripts) and [Turbolinks](https://github.com/turbolinks/turbolinks), which are both written in CoffeeScript and more-or-less use vanilla Sprockets. If this library is to be maintained by the Rails team then my feeling is we should follow those conventions.

Also, there was some discussion (https://github.com/rails/rails/issues/25208#issuecomment-222701608) by @simi and @sgrif about making this a GSoC (Google Summer of Code) project. Does anyone know what the status of that is?

**Comment by liudangyi** [comment]
@javan IMHO maybe that's because UJS and Action Cable are two different kinds of library. UJS is a collection of many functions. It does a lot of trivial things, like popping up a window when user clicks a button. Its functionality is wide but simple. However, Action Cable only focuses on one thing: provide an easy-to-use WebSocket API to work together with backend Rails. Its functionality is narrow and complex. So on code level, it's straightforward to write Action Cable in OOP paradigm, because there **is** something that needs to be managed. But if using OOP to write UJS, there would be a lot of classes with only one or two methods (such as `Ajax#send`, `DataRemoteHandler#handle`), which only adds complexity. Turbolinks is the same because it also does only one thing.

Another possible explanation why UJS looks so different may be that it's not written in CoffeeScript. Personally I have no bias on ES6 or CoffeeScript. But since original UJS was written in ES4 and I was just polishing the code rather than rewriting from scratch, I chose ES6. 

As for GSoC, I'm right the student enrolled to finish this project. Some information is available [here](https://summerofcode.withgoogle.com/projects/#6090835082870784) and [here (proposal)](https://docs.google.com/document/d/1yePH4Un-yW_rIh8ytqLQYSWc7nIuZMAeUmRF1qqkVbE/edit?usp=sharing). And I would like to continue maintaining this project after GSoC ends.

**Comment by liudangyi** [comment]
> Should it replace the `jquery-ujs` repo or have a new one for this vanilla js implementation?

I think `rails-ujs` should replace `jquery-ujs` because `jquery-ujs` is no longer to be developed any more. And `rails-ujs` is a comprehensive replacement for `jquery-ujs` because it has all the functions `jquery-ujs` has. 

We can consider `rails-ujs` as a new version of `jquery-ujs` (with an upgrade guide 😄)

**Comment by javan** [comment]
@liudangyi, Cool! I didn't know the GSoC project was happening. Thanks for letting me know.

>  But if using OOP to write UJS, there would be a lot of classes with only one or two methods

I'm not suggesting that Turbolinks or Action Cable's implementation be followed exactly. Just that they could serve as a guide for the file structure and build tooling.

> Turbolinks is the same because it also does only one thing.

Turbolinks does a *lot* of things: [event handling](https://github.com/turbolinks/turbolinks/blob/master/src/turbolinks/controller.coffee#L135-L152), [history management](https://github.com/turbolinks/turbolinks/blob/master/src/turbolinks/history.coffee), [DOM manipulations](https://github.com/turbolinks/turbolinks/blob/master/src/turbolinks/snapshot_renderer.coffee), and much more. Give the code a closer look if you haven't already.

> But since original UJS was written in ES4 and I was just polishing the code rather than rewriting from scratch, I chose ES6.

That makes sense and seems like a perfectly reasonable place to start. I imagine `jquery-ujs` has acquired quite a bit of baggage over the years, and I think it's worth considering what a ground up rewrite would look like.

Using ES6 + [sprockets-commoner](https://github.com/Shopify/sprockets-commoner) is only *slightly* different than using webpack directly as a build tool since [it requires `npm` and `babel-core`](https://github.com/Shopify/sprockets-commoner#10-second-simple-set-up). For that reason, it still feels very different than our other JavaScript libraries.

I don't want to dictate how this project is written or managed. Nor do I feel qualified to. Just sharing my thoughts!

**Comment by javan** [comment]
FWIW, I suspect the global `window.rails` object can be removed since everything is event driven and there's no public API. If we're going to call it unobtrusive then I think it should live up to its name.

**Comment by liudangyi** [comment]
> FWIW, I suspect the global window.rails object can be removed since everything is event driven and there's no public API. If we're going to call it unobtrusive then I think it should live up to its name.

Some test cases depend on this API. So currently we cannot remove it.

**Comment by javan** [comment]
Again, that can be considered baggage. I'm imagine the tests could be rewritten to work without a global or perhaps the global can be set in the test environment only.

Getting the original tests to pass seems like the best way to rewrite the implementation. Phase two could be improving that implementation overall.

**Comment by liudangyi** [comment]
> Again, that can be considered baggage. I'm imagine the tests could be rewritten to work without a global or perhaps the global can be set in the test environment only.

Yes, my next task is to rewrite the test to make it compatible with QUnit 1.20.0 and blade.

**Comment by dhh** [comment]
I concur with Javan on this. Going vanilla JS on the UJS adapter should not require us to diverge from the basic Sprockets setup we have. All the hooks are invoked by picking up on data- attributes anyway. It's not intended that the majority use case is someone calling our JS code for this directly. Let's stick with the precedence set by Action Cable and Turbolinks.

**Comment by liudangyi** [comment]
Another usage of `window.rails` is serving as a guard to prevent loading UJS twice. 

Personally I'm still in favor of JavaScript module style but keeping a consistency between different libraries seems more important. I'll first try to move from ES6 modules to Sprockets modules, then transform ES6 to CoffeeScript.

**Comment by liudangyi** [comment]
The code has been rewritten using CoffeeScript now. Could anyone help to review it?

**Comment by pixeltrix** [comment]
@liudangyi thanks your work on this - I'm not the biggest user of CoffeeScript so it's probably better that @javan is the final arbiter on that but it looks good to me. Also apologies on leading you astray with ES6 and WebPack - I think it's better in the end that we're consistent across all our JS projects.

@rafaelfranca is the plan to keep this code in the same repo or should we create a new repo for it? In some ways keeping it in the same repo makes sense - the legacy jquery-ujs would be on a separate branch and this new code would be our supported path going forward. My main concern is all of the existing links to this repo - if we rename it to `rails-ujs` how long does GitHub maintain redirects?

**Comment by rafaelfranca** [comment]
> @rafaelfranca is the plan to keep this code in the same repo or should we create a new repo for it? In some ways keeping it in the same repo makes sense - the legacy jquery-ujs would be on a separate branch and this new code would be our supported path going forward. My main concern is all of the existing links to this repo - if we rename it to rails-ujs how long does GitHub maintain redirects?

I could not find any information about redirects in the GitHub help apart from the note that if you create a repository with the same name the redirect will break.

I believe a new repository will be better for this case, there are still applications that will request jquery-ujs for a long time and having it as a branch here will make harder to deal with issues and pull requests since the reporter will have to point to the right branch.

**Comment by javan** [comment]
@liudangyi, at a glance it looks pretty good! I'll try to review more thoroughly soon. Have you tested it in an actual Rails app?

All, is the aim to exactly match the current behavior (but without jQuery) or should we take this opportunity review and maybe trim some fat?

**Comment by liudangyi** [comment]
@javan I was trying to keep the same behavior as old `jquery-ujs` and all old tests are kept and passed (with necessary modification). But I don't think we must comply with the old behavior, since there are already some differences in events and ajax now.

**Comment by connorshea** [comment]
I don't mean to act a peanut gallery or anything, but I prefer plain JavaScript over CoffeeScript if only because the general pattern has been moving to either Vanilla JavaScript or ES6.

We've moved from CS to ES6 at GitLab [1], and Atom is slowly moving all their CoffeeScript over to ES6 [2]. There's a lot of momentum in moving from CoffeeScript to vanilla JavaScript (at least, vanilla once ES6 becomes supported by all major browsers in the future), and I wanted to at least mention it before we dig ourselves into a deeper hole.

Also worth noting that [the CoffeeScript repository](https://github.com/jashkenas/coffeescript) hasn't seen as much activity lately, and `async`/`await` support still isn't there.

Anecdotal evidence, but after moving to ES6 at GitLab we've seen more frontend developers contributing to the project. CS just isn't as friendly to frontend developers when compared to vanilla/close-to-vanilla JavaScript.

Again, apologies for derailing this at all, and if there's a better place for me to open a discussion on this topic I'll gladly take it there.

[1]: https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/5464
[2]: https://github.com/atom/toggle-quotes/pull/26#issuecomment-157341949 https://github.com/atom/atom/commit/ab9a2a92c860007357f412935ff7bbbf2613af38

**Comment by gitmihalis** [comment]
Let me know if you'd like some help shoring this one up @liudangyi .

**Comment by guilleiguaran** [comment]
@liudangyi is this ready for review?

**Comment by liudangyi** [comment]
@guilleiguaran Yes.

File: Gemfile [hunk 1]
```diff
@@ -1,6 +1,10 @@
+$:.unshift 'lib'
 source 'https://rubygems.org'
 
 gem 'sinatra', '~> 1.0'
 gem 'shotgun', :group => :reloadable
 gem 'thin', :group => :reloadable
 gem 'rake'
+gem 'blade'
+gem 'uglifier'
+gem 'sprockets-commoner'
```
Comment (guilleiguaran) [inline Gemfile]: I don't think this is being used anymore in the PR, is it?

**Comment by guilleiguaran** [comment]
👍 on my side but would like to have the review of one of the maintainers of the library before of merging.

**Comment by rafaelfranca** [comment]
I still think this should be in another repository. jquery_ujs should be still here for people that use jquery and want to keep using. Also issues for jquery_ujs not necessarily will be issues in non-jquery_ujs, so even for issue triage that will be better.

**Comment by guilleiguaran** [comment]
@rafaelfranca agree totally, let's do it in that way

**Comment by guilleiguaran** [comment]
Ok, now the code of this PR is living under https://github.com/rails/rails-ujs, feel free to open PRs with improvements or open a issue in the new repo if you found any problem related to this.

**Comment by jakeNiemiec** [comment]
@guilleiguaran Out of curiosity, what happened to rails-ujs?

**Comment by guilleiguaran** [comment]
`rails-ujs` was merged inside of rails core: https://github.com/rails/rails/tree/master/actionview/app/assets/javascripts

**Comment by askehansen** [comment]
Thank you!
Spark by Readdle

**Pull request closed without merging**